File: 103.Binary Tree Zigzag Level Order Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    const stack = [root];
    const result = [];
    let count = 1;
    let current = 0;
    let depth = 0;
    while(stack.length) {
        const node = stack.shift();
        if (node) {
            if (current === 0) {
                result.push([]);
            }
            result[depth][depth % 2 === 0 ? 'push' : 'unshift'](node.val);
            if (node.left) {
                stack.push(node.left);
            }
            if (node.right) {
                stack.push(node.right);
            }
            current++;
            if (current === count) {
                current = 0;
                count = stack.length;
                depth++;
            }
        }
    }
    return result;
};

File: Golang/Tree/103.Binary-tree-zigzag-level-order-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func zigzagLevelOrder(root *TreeNode) [][]int {
	ret := [][]int{}
	if root != nil {
		stack := []*TreeNode{root}
		ltr := true
		for len(stack) > 0 {
			children := []*TreeNode{}
			temp := []int{}
			for i := 0; i < len(stack); i++ {
				node := stack[i]
				temp = append(temp, node.Val)
				if node.Left != nil {
					children = append(children, node.Left)
				}
				if node.Right != nil {
					children = append(children, node.Right)
				}
			}
			stack = children
			if !ltr {
				for i, j := 0, len(temp)-1; i < j; {
					temp[i], temp[j] = temp[j], temp[i]
					i++
					j--
				}
			}
			ret = append(ret, temp)
			ltr = !ltr
		}
	}
	return ret
}
